File: git-client/src/stores/remote.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { RemoteInfo, FetchResult, PullResult, PushResult } from '../types/git'
import { invoke } from '../utils/ipc'

export const useRemoteStore = defineStore('remote', () => {
  const remotes = ref<RemoteInfo[]>([])
  const loading = ref(false)
  const syncing = ref(false)

  async function fetchRemotes(repoPath: string) {
    loading.value = true
    try {
      remotes.value = await invoke<RemoteInfo[]>('list_remotes', { repoPath })
    } catch (e) {
      console.error('fetchRemotes error:', e)
    } finally {
      loading.value = false
    }
  }

  async function addRemote(repoPath: string, name: string, url: string) {
    await invoke('add_remote', { repoPath, name, url })
    await fetchRemotes(repoPath)
  }

  async function fetchRemote(repoPath: string, remote: string) {
    syncing.value = true
    try {
      return await invoke<FetchResult>('fetch', { repoPath, remote })
    } finally {
      syncing.value = false
    }
  }

  async function pullRemote(repoPath: string, remote: string, branch: string) {
    syncing.value = true
    try {
      return await invoke<PullResult>('pull', { repoPath, remote, branch })
    } finally {
      syncing.value = false
    }
  }

  async function pushRemote(repoPath: string, remote: string, branch: string) {
    syncing.value = true
    try {
      return await invoke<PushResult>('push', { repoPath, remote, branch })
    } finally {
      syncing.value = false
    }
  }

  return { remotes, loading, syncing, fetchRemotes, addRemote, fetchRemote, pullRemote, pushRemote }
})

File: git-client/src-tauri/src/commands/remote.rs
use crate::models::remote::RemoteInfo;
use crate::services::remote_service;
use crate::utils::error::AppError;
use crate::AppState;
use tauri::State;

#[tauri::command]
pub async fn list_remotes(
    state: State<'_, AppState>,
    repo_path: String,
) -> Result<Vec<RemoteInfo>, AppError> {
    let repos = state.repos.clone();
    tokio::task::spawn_blocking(move || {
        let manager = repos.lock().map_err(|e| AppError::Credential(e.to_string()))?;
        let repo = manager.get_repo(&repo_path)?;
        remote_service::list_remotes(&repo)
    })
    .await?
}

#[tauri::command]
pub async fn add_remote(
    state: State<'_, AppState>,
    repo_path: String,
    name: String,
    url: String,
) -> Result<(), AppError> {
    let repos = state.repos.clone();
    tokio::task::spawn_blocking(move || {
        let manager = repos.lock().map_err(|e| AppError::Credential(e.to_string()))?;
        let repo = manager.get_repo(&repo_path)?;
        remote_service::add_remote(&repo, &name, &url)
    })
    .await?
}

#[tauri::command]
pub async fn fetch(
    state: State<'_, AppState>,
    repo_path: String,
    remote: String,
) -> Result<remote_service::FetchResult, AppError> {
    let repos = state.repos.clone();
    tokio::task::spawn_blocking(move || {
        let manager = repos.lock().map_err(|e| AppError::Credential(e.to_string()))?;
        let repo = manager.get_repo(&repo_path)?;
        remote_service::fetch(&repo, &remote)
    })
    .await?
}

#[tauri::command]
pub async fn pull(
    state: State<'_, AppState>,
    repo_path: String,
    remote: String,
    branch: String,
) -> Result<remote_service::PullResult, AppError> {
    let repos = state.repos.clone();
    tokio::task::spawn_blocking(move || {
        let mut manager = repos.lock().map_err(|e| AppError::Credential(e.to_string()))?;
        let mut repo = manager.get_repo(&repo_path)?;
        remote_service::pull(&mut repo, &remote, &branch)
    })
    .await?
}

#[tauri::command]
pub async fn push(
    state: State<'_, AppState>,
    repo_path: String,
    remote: String,
    branch: String,
) -> Result<remote_service::PushResult, AppError> {
    let repos = state.repos.clone();
    tokio::task::spawn_blocking(move || {
        let manager = repos.lock().map_err(|e| AppError::Credential(e.to_string()))?;
        let repo = manager.get_repo(&repo_path)?;
        remote_service::push(&repo, &remote, &branch)
    })
    .await?
}
